File: library/src/main/java/su/levenetc/android/textsurface/animations/Rotate3D.java
package su.levenetc.android.textsurface.animations;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.PropertyValuesHolder;
import android.animation.ValueAnimator;
import android.graphics.Camera;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.view.animation.FastOutSlowInInterpolator;

import su.levenetc.android.textsurface.Text;
import su.levenetc.android.textsurface.TextSurface;
import su.levenetc.android.textsurface.utils.Utils;
import su.levenetc.android.textsurface.contants.Axis;
import su.levenetc.android.textsurface.contants.Pivot;
import su.levenetc.android.textsurface.interfaces.IEndListener;
import su.levenetc.android.textsurface.interfaces.ISurfaceAnimation;
import su.levenetc.android.textsurface.interfaces.ITextEffect;

/**
 * Created by Eugene Levenetc.
 */
public class Rotate3D implements ITextEffect, ValueAnimator.AnimatorUpdateListener {

	private Text text;
	private int duration;
	private int pivot;
	protected final Camera camera = new Camera();
	protected final Matrix cameraMatrix = new Matrix();
	private float rotationX;
	private float rotationY;
	private TextSurface textSurface;
	private float cameraTransXPre;
	private float cameraTransYPre;
	private float cameraTransXPost;
	private float cameraTransYPost;
	private int direction;
	private int axis;
	private boolean show;
	private ObjectAnimator animator;

	public static Rotate3D showFromSide(Text text, int duration, int pivot) {
		return new Rotate3D(text, duration, pivot, 0, 0, true);
	}

	public static Rotate3D showFromCenter(Text text, int duration, int direction, int axis) {
		return new Rotate3D(text, duration, Pivot.CENTER, direction, axis, true);
	}

	public static Rotate3D hideFromSide(Text text, int duration, int pivot) {
		return new Rotate3D(text, duration, pivot, 0, 0, false);
	}

	public static Rotate3D hideFromCenter(Text text, int duration, int direction, int axis) {
		return new Rotate3D(text, duration, Pivot.CENTER, direction, axis, false);
	}

	private Rotate3D(
			Text text,
			int duration,
			int pivot,
			int direction,
			int axis,
			boolean show
	) {
		this.text = text;
		this.duration = duration;
		this.pivot = pivot;
		this.direction = direction;
		this.axis = axis;
		this.show = show;
	}

	@Override public void apply(Canvas canvas, String textValue, float x, float y, Paint paint) {

		camera.getMatrix(cameraMatrix);
		camera.save();
		camera.rotateX(rotationX);
		camera.rotateY(rotationY);
		camera.getMatrix(cameraMatrix);

		cameraMatrix.preTranslate(cameraTransXPre, cameraTransYPre);
		cameraMatrix.postTranslate(cameraTransXPost, cameraTransYPost);

		camera.restore();

		canvas.concat(cameraMatrix);
	}

	@Override public void setInitValues(@NonNull Text text) {
		if (show) {
			text.setAlpha(0);
		}
	}

	@NonNull @Override public Text getText() {
		return text;
	}

	@Override public void onStart() {
		text.addEffect(this);
	}

	@Override public void start(@Nullable final IEndListener listener) {


		PropertyValuesHolder valHolder = null;

		int fromDegree;
		int toDegree;

		text.setAlpha(255);

		if (show) {
			fromDegree = 90;
			toDegree = 0;
		} else {
			fromDegree = 0;
			toDegree = -90;
		}

		if ((pivot & Pivot.BOTTOM) == Pivot.BOTTOM) {

			valHolder = PropertyValuesHolder.ofFloat("rotationX", fromDegree, toDegree);
			cameraTransXPre = -text.getWidth() / 2;
			cameraTransXPost = text.getWidth() / 2;
			cameraTransYPre = -text.getFontDescent();
			cameraTransYPost = 0;
		} else if ((pivot & Pivot.TOP) == Pivot.TOP) {
			valHolder = PropertyValuesHolder.ofFloat("rotationX", -fromDegree, toDegree);
			cameraTransXPre = -text.getWidth() / 2;
			cameraTransXPost = text.getWidth() / 2;
			cameraTransYPre = text.getHeight() - text.getFontDescent();
			cameraTransYPost = -text.getHeight();
		}

		if ((pivot & Pivot.LEFT) == Pivot.LEFT) {
			valHolder = PropertyValuesHolder.ofFloat("rotationY", fromDegree, toDegree);
			cameraTransXPre = 0;
			cameraTransXPost = 0;
			cameraTransYPre = text.getHeight() / 2 - text.getFontDescent();
			cameraTransYPost = text.getHeight() / 2 - text.getHeight();
		} else if ((pivot & Pivot.RIGHT) == Pivot.RIGHT) {
			valHolder = PropertyValuesHolder.ofFloat("rotationY", -fromDegree, toDegree);
			cameraTransXPre = -text.getWidth();
			cameraTransXPost = text.getWidth();
			cameraTransYPre = text.getHeight() / 2 - text.getFontDescent();
			cameraTransYPost = text.getHeight() / 2 - text.getHeight();
		}

		if ((pivot & Pivot.CENTER) == Pivot.CENTER) {
			valHolder = PropertyValuesHolder.ofFloat(axis == Axis.Y ? "rotationY" : "rotationX", fromDegree, toDegree);
			cameraTransXPre = -text.getWidth() / 2;
			cameraTransXPost = text.getWidth() / 2;
			cameraTransYPre = text.getHeight() / 2 - text.getFontDescent();
			cameraTransYPost = text.getHeight() / 2 - text.getHeight();
		}

		if (valHolder != null) {
			animator = ObjectAnimator.ofPropertyValuesHolder(this, valHolder);
			animator.setInterpolator(new FastOutSlowInInterpolator());
			Utils.addEndListener(this, animator, new IEndListener() {
				@Override public void onAnimationEnd(ISurfaceAnimation animation) {
					text.removeEffect(Rotate3D.this);
					if (!show) text.setAlpha(0);
					if (listener != null) listener.onAnimationEnd(Rotate3D.this);
				}
			});
			animator.setDuration(duration);
			animator.addUpdateListener(this);
			animator.start();

		} else {
			throw new RuntimeException(getClass().getSuperclass() + " was not configured properly. Pivot:" + pivot);
		}

	}

	@Override public void setTextSurface(@NonNull TextSurface textSurface) {

		this.textSurface = textSurface;
	}

	@Override public long getDuration() {
		return duration;
	}

	@Override public void cancel() {
		if (animator != null && animator.isRunning()) {
			animator.cancel();
			animator = null;
		}
	}

	public void setRotationX(float rotationX) {
		this.rotationX = rotationX;
	}

	public void setRotationY(float rotationY) {
		this.rotationY = rotationY;
	}

	@Override public void onAnimationUpdate(ValueAnimator animation) {
		textSurface.invalidate();
	}
}

File: library/src/main/java/su/levenetc/android/textsurface/animations/Rotate3D.kt
package su.levenetc.android.textsurface.animations

import android.animation.ObjectAnimator
import android.animation.PropertyValuesHolder
import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.graphics.*
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import su.levenetc.android.textsurface.Text
import su.levenetc.android.textsurface.TextSurface
import su.levenetc.android.textsurface.contants.Axis
import su.levenetc.android.textsurface.contants.Pivot
import su.levenetc.android.textsurface.interfaces.IEndListener
import su.levenetc.android.textsurface.interfaces.ISurfaceAnimation
import su.levenetc.android.textsurface.interfaces.ITextEffect
import su.levenetc.android.textsurface.utils.Utils.addEndListener

/**
 * Created by Eugene Levenetc.
 */
class Rotate3D private constructor(
        override val text: Text,
        override var duration: Long,
        private val pivot: Int,
        private val direction: Int,
        private val axis: Int,
        private val show: Boolean
) : ITextEffect, AnimatorUpdateListener {
    protected val camera = Camera()
    protected val cameraMatrix = Matrix()
    private var rotationX = 0f
    private var rotationY = 0f
    override lateinit var textSurface: TextSurface
    private var cameraTransXPre = 0f
    private var cameraTransYPre = 0f
    private var cameraTransXPost = 0f
    private var cameraTransYPost = 0f
    private var animator: ObjectAnimator? = null

    override fun apply(canvas: Canvas, textValue: String, x: Float, y: Float, paint: Paint) {
        camera.getMatrix(cameraMatrix)
        camera.save()
        camera.rotateX(rotationX)
        camera.rotateY(rotationY)
        camera.getMatrix(cameraMatrix)
        cameraMatrix.preTranslate(cameraTransXPre, cameraTransYPre)
        cameraMatrix.postTranslate(cameraTransXPost, cameraTransYPost)
        camera.restore()
        canvas.concat(cameraMatrix)
    }

    override fun setInitValues(text: Text) {
        if (show) {
            text.setAlpha(0)
        }
    }

    override fun onStart() {
        text.addEffect(this)
    }

    override fun start(listener: IEndListener) {
        var valHolder: PropertyValuesHolder? = null
        val fromDegree: Int
        val toDegree: Int
        text.setAlpha(255)
        if (show) {
            fromDegree = 90
            toDegree = 0
        } else {
            fromDegree = 0
            toDegree = -90
        }
        if (pivot and Pivot.BOTTOM == Pivot.BOTTOM) {
            valHolder = PropertyValuesHolder.ofFloat("rotationX", fromDegree.toFloat(), toDegree.toFloat())
            cameraTransXPre = -text.width / 2
            cameraTransXPost = text.width / 2
            cameraTransYPre = -text.fontDescent
            cameraTransYPost = 0f
        } else if (pivot and Pivot.TOP == Pivot.TOP) {
            valHolder = PropertyValuesHolder.ofFloat("rotationX", -fromDegree.toFloat(), toDegree.toFloat())
            cameraTransXPre = -text.width / 2
            cameraTransXPost = text.width / 2
            cameraTransYPre = text.height - text.fontDescent
            cameraTransYPost = -text.height
        }
        if (pivot and Pivot.LEFT == Pivot.LEFT) {
            valHolder = PropertyValuesHolder.ofFloat("rotationY", fromDegree.toFloat(), toDegree.toFloat())
            cameraTransXPre = 0f
            cameraTransXPost = 0f
            cameraTransYPre = text.height / 2 - text.fontDescent
            cameraTransYPost = text.height / 2 - text.height
        } else if (pivot and Pivot.RIGHT == Pivot.RIGHT) {
            valHolder = PropertyValuesHolder.ofFloat("rotationY", -fromDegree.toFloat(), toDegree.toFloat())
            cameraTransXPre = -text.width
            cameraTransXPost = text.width
            cameraTransYPre = text.height / 2 - text.fontDescent
            cameraTransYPost = text.height / 2 - text.height
        }
        if (pivot and Pivot.CENTER == Pivot.CENTER) {
            valHolder = PropertyValuesHolder.ofFloat(if (axis == Axis.Y) "rotationY" else "rotationX", fromDegree.toFloat(), toDegree.toFloat())
            cameraTransXPre = -text.width / 2
            cameraTransXPost = text.width / 2
            cameraTransYPre = text.height / 2 - text.fontDescent
            cameraTransYPost = text.height / 2 - text.height
        }
        if (valHolder != null) {
            animator = ObjectAnimator.ofPropertyValuesHolder(this, valHolder)
            animator!!.interpolator = FastOutSlowInInterpolator()
            addEndListener(this, animator!!, object : IEndListener {
                override fun onAnimationEnd(animation: ISurfaceAnimation) {
                    text.removeEffect(this@Rotate3D)
                    if (!show) text.setAlpha(0)
                    listener.onAnimationEnd(this@Rotate3D)
                }
            })
            animator!!.duration = duration.toLong()
            animator!!.addUpdateListener(this)
            animator!!.start()
        } else {
            throw RuntimeException(javaClass.superclass.toString() + " was not configured properly. Pivot:" + pivot)
        }
    }

    override fun cancel() {
        if (animator != null && animator!!.isRunning) {
            animator!!.cancel()
            animator = null
        }
    }

    fun setRotationX(rotationX: Float) {
        this.rotationX = rotationX
    }

    fun setRotationY(rotationY: Float) {
        this.rotationY = rotationY
    }

    override fun onAnimationUpdate(animation: ValueAnimator) {
        textSurface!!.invalidate()
    }

    companion object {
        @kotlin.jvm.JvmStatic
        fun showFromSide(text: Text, duration: Long, pivot: Int): Rotate3D {
            return Rotate3D(text, duration, pivot, 0, 0, true)
        }

        @kotlin.jvm.JvmStatic
        fun showFromCenter(text: Text, duration: Long, direction: Int, axis: Int): Rotate3D {
            return Rotate3D(text, duration, Pivot.CENTER, direction, axis, true)
        }

        @kotlin.jvm.JvmStatic
        fun hideFromSide(text: Text, duration: Long, pivot: Int): Rotate3D {
            return Rotate3D(text, duration, pivot, 0, 0, false)
        }

        @kotlin.jvm.JvmStatic
        fun hideFromCenter(text: Text, duration: Long, direction: Int, axis: Int): Rotate3D {
            return Rotate3D(text, duration, Pivot.CENTER, direction, axis, false)
        }
    }
}